src/common.c

```diff
@@ -707,60 +709,191 @@ int fi_epoll_create(struct fi_epoll **ep)
 {
+	struct slist_entry *entry, *prev;
+	struct fi_epoll_work_item *item, *del_item;
 	int i;
 
-	for (i = 0; i < ep->nfds; i++) {
+	del_item = calloc(1,sizeof(*item));
+	if (!del_item)
+		return -FI_ENOMEM;
+
+	del_item->fd = fd;
+	del_item->type = FI_EPOLL_DEL;
+
+	fastlock_acquire(&ep->lock);
+	slist_foreach(&ep->work_item_list, entry, prev) {
+		item = container_of(entry, struct fi_epoll_work_item,
+				    entry);
+		if (item->fd == fd &&
+		    (item->type == FI_EPOLL_ADD)) {
+			goto out;
+		}
+	}
+
+	for (i = 1; i < ep->nfds; i++) {
 		if (ep->fds[i].fd == fd) {
-			ep->fds[i].fd = ep->fds[ep->nfds - 1].fd;
-			ep->context[i] = ep->context[--ep->nfds];
-			fd_signal_set(&ep->signal);
-			return 0;
+			goto out;
 		}
   	}
+	fastlock_release(&ep->lock);
+	free(del_item);
 	return -FI_EINVAL;
+out:
+	slist_insert_tail(&del_item->entry, &ep->work_item_list);
+	fd_signal_set(&ep->signal);
+	fastlock_release(&ep->lock);
+	return 0;
+}
+
+static int fi_epoll_fd_array_grow(struct fi_epoll *ep)
+{
+	struct pollfd *fds;
+	void *contexts;
+
+	fds = calloc(ep->size + 64,
+		     sizeof(*ep->fds) + sizeof(*ep->context));
+	if (!fds)
+		return -FI_ENOMEM;
+
+	ep->size += 64;
+	contexts = fds + ep->size;
+
+	memcpy(fds, ep->fds, ep->nfds * sizeof(*ep->fds));
+	memcpy(contexts, ep->context, ep->nfds * sizeof(*ep->context));
+	free(ep->fds);
+	ep->fds = fds;
+	ep->context = contexts;
+	return FI_SUCCESS;
+}
+
+static void fi_epoll_cleanup_array(struct fi_epoll *ep)
+{
+	int i;
+
+	for (i = 0; i < ep->nfds; i++) {
+		while (ep->fds[i].fd < 0) {
+			ep->fds[i].fd = ep->fds[ep->nfds-1].fd;
+			ep->fds[i].events = ep->fds[ep->nfds-1].events;
+			ep->fds[i].revents = ep->fds[ep->nfds-1].revents;
+			ep->context[i] = ep->context[ep->nfds-1];
+			ep->nfds--;
+			if (i == ep->nfds)
+				break;
+		}
+	}
+}
+
+static int fi_epoll_process_work_item_list(struct fi_epoll *ep)
+{
+	struct slist_entry *entry;
+	struct fi_epoll_work_item *item;
+	int i, ret = 0;
+
+	while (!slist_empty(&ep->work_item_list)) {
+		if ((ep->nfds == ep->size) &&
+		    fi_epoll_fd_array_grow(ep))
+			continue;
+
+		entry = slist_remove_head(&ep->work_item_list);
+		item = container_of(entry, struct fi_epoll_work_item, entry);
+
+		switch (item->type) {
+		case FI_EPOLL_ADD:
+			ep->fds[ep->nfds].fd = item->fd;
+			ep->fds[ep->nfds].events = item->events;
+			ep->context[ep->nfds] = item->context;
+			ep->nfds++;
+			break;
+		case FI_EPOLL_DEL:
+			for (i = 0; i < ep->nfds; i++) {
+				if (ep->fds[i].fd == item->fd) {
+					ep->fds[i].fd = -1;
+					break;
+				}
+			}
+			break;
+		case FI_EPOLL_MOD:
+			for (i = 0; i < ep->nfds; i++) {
+				if (ep->fds[i].fd == item->fd) {
+
+					ep->fds[i].events = item->events;
+					ep->fds[i].revents &= item->events;
+					ep->context = item->context;
+					break;
+				}
+			}
+			break;
+		default:
+			ret = -FI_EINVAL;
+			goto out;
+		}
+		free(item);
+	}
+out:
+	fi_epoll_cleanup_array(ep);
+	return ret;
```

Comment:
I'd drop the return statement from this function and ignore errors, since we can't do anything with them anyway at this point.